FreeRTOS/Test/CMock/message_buffer/message_buffer_utest.c

```diff
@@ -0,0 +1,670 @@
+void tearDown( void )
+{
+    TEST_ASSERT_EQUAL_MESSAGE( 0, assertionFailed, "Assertion check failed in code." );
+    UnityMalloc_EndTest();
+}
+
+/*! called at the beginning of the whole suite */
+void suiteSetUp()
+{
+}
+
+/*! called at the end of the whole suite */
+int suiteTearDown( int numFailures )
+{
+    return numFailures;
+}
+
+static void validate_message_buffer_init_state( MessageBufferHandle_t xMessageBuffer,
+                                                size_t bufferSize )
+{
+    TEST_ASSERT_TRUE( xMessageBufferIsEmpty( xMessageBuffer ) );
+    TEST_ASSERT_FALSE( xMessageBufferIsFull( xMessageBuffer ) );
+    TEST_ASSERT_EQUAL( bufferSize, xMessageBufferSpacesAvailable( xMessageBuffer ) );
+    TEST_ASSERT_EQUAL( 0, xStreamBufferNextMessageLengthBytes( xMessageBuffer ) );
+    TEST_ASSERT_EQUAL( 1, ucStreamBufferGetStreamBufferType( xMessageBuffer ) );
+}
+
+static void validate_and_clear_assertitions( void )
+{
+    TEST_ASSERT_EQUAL( 1, assertionFailed );
+    assertionFailed = 0;
+}
+
+/* ==============================  Test Cases  ============================== */
+
+/**
+ * @brief Validates that message buffer of sample test size is created successfully.
+ */
+void test_xMessageBufferCreate_success( void )
+{
+    xMessageBuffer = xMessageBufferCreate( TEST_MESSAGE_BUFFER_SIZE );
+    TEST_ASSERT_NOT_EQUAL( NULL, xMessageBuffer );
+    validate_message_buffer_init_state( xMessageBuffer, TEST_MESSAGE_BUFFER_SIZE );
+    vMessageBufferDelete( xMessageBuffer );
+}
+
+/**
+ * @brief Validates message buffer create fails if invalid parameters are passed.
+ */
+void test_xMessageBufferCreate_invalid_params( void )
+{
+    /* API should fail and return null if there is an integer overflow in message buffer size passed. */
+    xMessageBuffer = xMessageBufferCreate( TEST_MESSAGE_BUFFER_MAX_UINT_SIZE );
+    TEST_ASSERT_EQUAL( NULL, xMessageBuffer );
+
+    /* API should fail and return null if the malloc fails. */
+    UnityMalloc_MakeMallocFailAfterCount( 0 );
+    xMessageBuffer = xMessageBufferCreate( TEST_MESSAGE_BUFFER_SIZE );
+    TEST_ASSERT_NULL( xMessageBuffer );
+
+    /* Should assert if a zero size is passed to create API. */
+    if( TEST_PROTECT() )
+    {
+        ( void ) xMessageBufferCreate( 0 );
+    }
+
+    validate_and_clear_assertitions();
+
+    /* Should assert if the size passed is less than the minimum size requried to store metadata size. */
+    if( TEST_PROTECT() )
+    {
+        ( void ) xMessageBufferCreate( TEST_MESSAGE_METADATA_SIZE );
+    }
+
+    validate_and_clear_assertitions();
+}
+
+/**
+ * @brief Validates message buffer creation through a static buffer allocated by the user.
+ */
+void test_xMessageBufferCreateStatic_success( void )
+{
+    StaticMessageBuffer_t messageBufferStruct = { 0 };
+    /* The size of message buffer array should be one greater than the required size of message buffer. */
+    uint8_t messageBufferArray[ TEST_MESSAGE_BUFFER_SIZE + 1 ] = { 0 };
```

Comment:
It would be safer to heap allocate your static buffers so that the address sanitizer can check correct access.

Reply:
As per docs, address sanitizer should catch buffer overflow on stack memory as well.
https://github.com/google/sanitizers/wiki/AddressSanitizerExampleStackOutOfBounds

Reply:
Yes. My bad. I confirmed that AddressSanitizer handles stack allocated variables as well on GCC 9.3.0.